scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedOpInfo.java

```diff
@@ -0,0 +1,366 @@
+	 * Computers.Arity1, would need a {@link Computers.Arity1} copy Op to copy the
+	 * output of the simplified Op back into the preallocated output. But this
+	 * call is already identical to the Op we asked for, and we know that there is
+	 * no direct match, thus we go again into simplification. This thus causes an
+	 * infinite loop (and eventually a {@link StackOverflowError}. This means that
+	 * we cannot find a simplified copy Op <b>unless a direct match can be
+	 * found</b>, at which point we might as well just use the direct match.
+	 * <p>
+	 * Adaptation is similarly forbidden, as to convert most Op types to
+	 * {@link Computers.Arity1} you would need an identical copy Op.
+	 * <p>
+	 *
+	 * @param env - the {@link OpEnvironment} containing the copy Ops
+	 * @param opType - the {@link FunctionalInterface} implemented by the Op.
+	 * @param args - the input {@link Type}s to the Op
+	 * @return an {@code Op} able to copy data between {@link Object}s of
+	 *         {@link Type} {@code copyType}
+	 */
+	private static RichOp<Computers.Arity1<?, ?>> simplifierCopyOp(
+		OpEnvironment env, Class<?> opType, List<Type> args)
+	{
+		int mutableIndex = SimplificationUtils.findMutableArgIndex(opType);
+		if (mutableIndex == -1) {
+			return null;
+		}
+		Type copyType = args.get(mutableIndex);
+		// prevent further simplification/adaptation
+		Hints hints = new Hints(BaseOpHints.Adaptation.FORBIDDEN,
+			BaseOpHints.Simplification.FORBIDDEN);
+		Nil<?> copyNil = Nil.of(copyType);
+		var op = env.unary("copy", hints).inType(copyNil).outType(copyNil)
+			.computer();
+		return Ops.rich(op);
+	}
+
+	/**
+	 * We define the priority of any {@link SimplifiedOpInfo} as the sum of the
+	 * following:
+	 * <ul>
+	 * <li>{@link Priority#VERY_LOW} to ensure that simplifications are not chosen
+	 * over a direct match.</li>
+	 * <li>The {@link OpInfo#priority} of the source info to ensure that a
+	 * simplification of a higher-priority Op wins out over a simplification of a
+	 * lower-priority Op, all else equal.</li>
+	 * <li>a penalty defined as a lossiness heuristic of this simplification. This
+	 * penalty is the sum of:</li>
+	 * <ul>
+	 * <li>the loss undertaken by converting each of the Op's inputs from the ref
+	 * type to the info type</li>
+	 * <li>the loss undertaken by converting each of the Op's outputs from the
+	 * info type to the ref type</li>
+	 * </ul>
+	 * </ul>
+	 */
+	private double calculatePriority(OpEnvironment env) {
+		// BASE PRIORITY
+		double base = Priority.VERY_LOW;
+
+		// ORIGINAL PRIORITY
+		double originalPriority = simpleInfo.srcInfo().priority();
+
+		// PENALTY
+		double penalty = 0;
+
+		List<Type> originalInputs = simpleInfo.srcInfo().inputTypes();
+		List<Type> inputs = inputTypes();
+		for (int i = 0; i < inputs.size(); i++) {
+			penalty += determineLoss(env, Nil.of(inputs.get(i)), Nil.of(originalInputs
+				.get(i)));
+		}
+
+		// TODO: only calculate the loss once
```

Comment:
Uh, I think the point was that we want to avoid computing the priority every time we call `FocusedOpInfo.getPriority()`, but I have no idea why the `TODO` was there.

Since there isn't a major harm in computing it more than once, I just added a `final Double priority = null` field to `FocusedOpInfo`, and whenever `FocusedOpInfo.getPriority()` is called we compute the priority if `priority == null`. I think that's enough to kill the `TODO`